x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/IndexServiceAccountsTokenStore.java

```diff
@@ -0,0 +1,196 @@
+                    listener.onResponse(Hasher.verifyHash(token.getSecret(), tokenHash.toCharArray()));
+                } else {
+                    logger.trace("service account token [{}] not found in index", token.getQualifiedName());
+                    listener.onResponse(false);
+                }}, listener::onFailure)));
+    }
+
+    public void createToken(Authentication authentication, CreateServiceAccountTokenRequest request,
+                            ActionListener<CreateServiceAccountTokenResponse> listener) {
+        final ServiceAccountId accountId = new ServiceAccountId(request.getNamespace(), request.getServiceName());
+        if (false == ServiceAccountService.isServiceAccountPrincipal(accountId.asPrincipal())) {
+            listener.onFailure(new IllegalArgumentException("service account [" + accountId + "] does not exist"));
+            return;
+        }
+        final ServiceAccountToken token = ServiceAccountToken.newToken(accountId, request.getTokenName());
+        try (XContentBuilder builder = newDocument(authentication, token)) {
+            final IndexRequest indexRequest =
+                client.prepareIndex(SECURITY_MAIN_ALIAS)
+                    .setId(docIdForToken(token))
+                    .setSource(builder)
+                    .setOpType(OpType.CREATE)
+                    .setRefreshPolicy(request.getRefreshPolicy())
+                    .request();
+            final BulkRequest bulkRequest = toSingleItemBulkRequest(indexRequest);
+
+            securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+                executeAsyncWithOrigin(client, SECURITY_ORIGIN, BulkAction.INSTANCE, bulkRequest,
+                    TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(response -> {
+                        if (DocWriteResponse.Result.CREATED == response.getResult()) {
+                            listener.onResponse(CreateServiceAccountTokenResponse.created(
+                                token.getTokenName(), token.asBearerString()));
+                        } else {
+                            listener.onResponse(CreateServiceAccountTokenResponse.notCreated());
+                        }
+                    }, listener::onFailure)));
+            });
+        } catch (IOException e) {
+            listener.onFailure(e);
+        }
+    }
+
+    @Override
+    public void findTokensFor(ServiceAccountId accountId, ActionListener<Collection<TokenInfo>> listener) {
+        // TODO: wildcard support?
+        final BoolQueryBuilder query = QueryBuilders.boolQuery()
+            .filter(QueryBuilders.termQuery("doc_type", SERVICE_ACCOUNT_TOKEN_DOC_TYPE))
+            .must(QueryBuilders.prefixQuery("name", accountId.asPrincipal()));
+        final SearchRequest request = client.prepareSearch(SECURITY_MAIN_ALIAS)
+            .setScroll(DEFAULT_KEEPALIVE_SETTING.get(getSettings()))
```

Comment:
Do we really want to scroll here, or should we place a limit on how many tokens there can be for a service account?

Reply:
I was going to say we should replace this and a few other places with the current best practice of "PIT + search_after". I need to check the details, but I think PIT is sharable across different requests. Given we need it in places, e.g. users, application privileges, it is not much overhead to have it here as well. It can be a follow up PR and I can raise an issue for it.
Now, it is also possible to cap it as you suggested. But I am not very convinced because:
1. It feels arbitrary. Yes it is unlikely to have many service tokens. But it does not inherently says otherwise either.
2. It introduces extra code complexity, e.g. you need query the current count before creating a new one. There can also be complicated concurrent issues, e.g. two requests at the same time, one tries to create and the other tries to delete. We can potential solve these issues, but the effort is not really justified.

Reply:
The problem is that scrolls are a limited resource, and we overuse them.
What we really want is semantics that only create a scroll (or PIT etc) if there were more documents than could be returned in a single request, but I don't know of anything like that.

A regular PIT could work, but 99% of the time it's heavier than we need.

Reply:
I wonder whether this could be a feature request for the search team. It is not possible to be perfect unless directly tackled on the shard level. Anyway, I think this worth to be its own separate task and can wait till the current one is over?

Reply:
Yes, I think it can wait. But I know @albertzaharovits ran into some cases recently where security was over-using scrolls to the detriment of the overall cluster, so I'm extra conscious of them now.

Reply:
After discussion, I raised an issue for replacing scroll with PIT + search_after https://github.com/elastic/elasticsearch/issues/71032